src/include/rados/librados.hpp

```diff
@@ -343,6 +345,20 @@ inline namespace v14_2_0 {
     void exec(const char *cls, const char *method, bufferlist& inbl);
     void exec(const char *cls, const char *method, bufferlist& inbl, bufferlist *obl, int *prval);
     void exec(const char *cls, const char *method, bufferlist& inbl, ObjectOperationCompletion *completion);
+   protected:
+    void exec_ro(const char *cls, const char *method, bufferlist& inbl);
+    void exec_ro(const char *cls, const char *method, bufferlist& inbl, bufferlist *obl, int *prval);
+    void exec_ro(const char *cls, const char *method, bufferlist& inbl, ObjectOperationCompletion *completion);
+   public:
+    template <typename Tag, typename ClassID, typename... Args>
+    void exec(const ClsMethod<Tag, ClassID>& method, Args&&... args) {
+      if constexpr (FlagTraits<Tag>::is_readonly) {
+        exec_ro(method.cls, method.name, std::forward<Args>(args)...);
+      } else {
+        exec(method.cls, method.name, std::forward<Args>(args)...);
+      }
```

Comment:
as-is, we wouldn't catch misuses like this:
```c++
void inc_version(ObjectOperation& op)
{
  bufferlist in;
  op.exec(::cls::version::MethodInc, in);
}

void foo()
{
  ObjectReadOperation op;
  inc_version(op);
```
probably better for `ObjectOperation::exec()` to only support reads, with a new `ObjectWriteOperation::exec()` that supports both reads and writes